bside-code/bside.1
.TH "bside" 1
.SH NAME
 BSide(Basic Stamp IDE)
.SH SYNOPSIS
.B bside 
[-vht?] [-d device] -f filename [-p program slot]
.SH DESCRIPTION
bside \- bside is a program for loading BASIC code onto a BASIC Stamp. bside supports all a number of different basic stamps including all 7 BASIC stamps by Parallax. It has the ability to pass custom program slots to the stamp. It is written in C and C++.
.SH OPTIONS
.TP
.B \-h or \-?
Print usage information. 
.TP
.B \-d
Specify the path to the desired device. The default device is '/dev/ttyUSB0' 
.TP
.B \-f
Specify the path to the file to load to the stamp.
.TP
.B \-v
Print bside version information.
.TP
.B \-p
Set custom program slot. As far as I know this is only used in Parallax's stamp.
.TP
.B \-t
Test Mode (Run whatever is on the stamp)
.TP
.SH NOTES
Basic Stamps from Parallax require the tokenizer library by Parallax, Inc. This tokenizer is a binary only shared object and non-free. A free version is planned by the bside project in the future. Please consider helping out!
.SH EXAMPLES
.TP
bside -d /dev/ttyS1 -f filename
.SH DEVICES
bside supports both usb and serial interfaces. Please see the included INSTALL text for information on device creation.
.SH LICENSE
bside is released under version two of the GPL. Please see the license file.
.SH SEE ALSO
Please see the included INSTALL and README files for more information.
.SH BUGS
Please send bug reports to Scott Sloan <devscott@charter.net> with your stamp vendor, stamp model, stamp firmware, type of interface and the steps to recreate your issue.
.SH AUTHORS
bside was written by the bside team led by Scott Sloan <devscott@charter.net>. This manual page was originally written by Jacob Appelbaum <jacob@appelbaum.net>.


bside-code/CHANGELOG
BSide Changelog

bside 0.05 September 25, 2006 ""
	* Cleaned up argument handling in bside_tokenizer (Katie - katie@hoteldetective.org)
	* General Code Clean up ( Scott - devscott@charter.net)
	* General Code Clean up ( Katie - katie@hoteldetective.org)
	*

bside 0.04 August 28, 2006 "A boy named Sue"
	* Fixed Bug where bside would hang if no stamp was connected (Scott - devscott@charter.net)
	* Fixed Bug where program would run on -? argument (Scott - devscott@charter.net)
	* Fixed Bug where Parallax Stamp wouldn't be closed after failed detection (Scott - devscott@charter.net)
	* Fixed Bug where Coridium Stamp would not be started (Scott - devscott@charter.net)
	* Fixed Possible Sigfault when invalid device is used (Scott - devscott@charter.net)
	* More Documentation Updates ( Scott - devscott@charter.net)
	* General Code Clean up (Scott - devscott@charter.net)
	* Pushed version 0.04 (Scott - devscott@charter.net)

bside 0.03  August 1, 2006 "Slow and Steady wins the race"
	* Added read support for Coridium's ARMexpress (Scott  - devscott@charter.net)
	* Updated man page to reflect the new support for Coridium Stamps (Scott  - devscott@charter.net)
	* Updated the README to reflect the new support for Coridium Stamps (Scott - devscott@charter.net)
	* Updated the README on USB Support (Scott - devscott@charter.net)
	* Updated the INSTALL on USB Support (Scott - devscott@charter.net)
	* Display Vender in Stamp info when detected ( Scott - devscott@charter.net)
	* Created Vender Specific headers (Scott - devscott@charter.net)
	* Moved all Vendor Specific code to it's header file (Scott - devscott@charter.net)
	* Updated Makefile (Scott - devscott@charter.net)
	* Fixed Bug where an invalid program slot could be set for Parallax stamps (Scott - devscott@charter.net)
	* Fixed Bug where device wouldn't close on error. (Scott - devscott@charter.net)
	* Fixed Bug where error messages were not being displayed (Scott - devscott@charter.net)

bside 0.02.2 "A Second Time" June 26, 2006
	* Fixed incorrect path for libbsidetokenizer (Katie - katie@hoteldetective.org)
	* Updated man page (Scott - devscott@charter.net)
	* Updated Install File (Adam - gnome@dux-linux.org)

bside 0.02 "A Spoon Full of Cherries": June 26, 2006
	* Created stamp.h stamp.cpp to handle stamp Model and firmware version 
	*	-> This is an initial step to become more OO
	*	(Scott - devscott@charter.net)

	* Added -t option for test mode (Scott - devscott@charter.net)
	* Added install file (Adam - gnome@dux-linux.org)
	* Added Man Page (Jacob - jacob@appelbaum.net)
	* Cleaned up stamp_sendFile() (Scott - devscott@charter.net)
	* Added programSlots count to stamp.h (Scott - devscott@charter.net)
	* Updated man page for test mode (Scott - devscott@charter.net)
	* Fixed incorrect path for libbsidetokenizer (Katie - katie@hoteldetective.org)

bside-0.01 "Happy Birthday":	June 15, 2006
	* initial Release


bside-code/bside.cpp
/*	Project Information:
 *		Name: BSide(Basic Stamp IDE)
 *		Page: http://www.devscott.org/projects/bside
 *		License: GPL
 * 		
 * 		About:
 */

#ifndef _GNU_SOURCE
# define _GNU_SOURCE 1		/* enables some compiler checks on GNU */
#endif

#define DEBUG 1

#include <iostream>
#include <string>
#include <fstream>
#include <cerrno>
#include <iomanip>

extern "C" {
#include <signal.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
}

#include "stamp.h"

using namespace std;

stamp myStamp;

/*General Stuff */

float bside_version = 0.04;		/* Version of Software */
string myFile = "";	    		/* File to be transfered */
string device;        			/* Stamp device */

/* Stamp Stuff */

int programSlot = 0;			/* Program Slot for Parallax stamps */
int fd;

/* Debug Stuff */
bool debug_loop_done;
bool testRun;

void alarmTimeout(int);
void *control_c_handler(int sig);
void stamp_debug(void);

static void help()
{
	printf("Usage: bside -d [device] -f [filename]\n\n");
	printf("Arguments:\n\n");
	printf("\t-d    --device\t Specify the desired device path\n");
	printf("\t-f    --file\t Specify file to send to the Stamp\n");
	printf("\t-h    --help\t This screen\n");						
	printf("\t-s    --slot\t Set the Program Slot (0-7) Parallax Stamps Only\n");
	printf("\t-t    --test\t Run in Test Mode\n");						
	printf("\t-v    --version\t Get version information\n");
	printf("\n");
}

static void version()
{
	printf("BSide Version: %f\n", bside_version);
}

int main(int argc, char **argv)
{	
	bool mode_test = false;
	int opt; 

	static struct option long_options[] = {
		{ "device", 	required_argument, 	0,	'd'},
		{ "file", 	required_argument, 	0,	'f'},
		{ "help", 	no_argument,		NULL, 	'h'},
		{ "slot", 	required_argument, 	0,	'p'},
		{ "test",	no_argument,		0,	't'},
		{ "version", 	no_argument, 		NULL, 	'v'},
		{0, 0, 0, 0}
	};


	if ( argc > 1)
	{
		while ((opt = getopt_long(argc, argv, ":th?s:vd:f:", long_options, NULL) )!= -1)
		{
			switch(opt)
			{
				case 'h': case '?':
					help();
					return 0;
					break;
				case 'd':
					myStamp.set_devicePort(strdup(optarg));     
					#ifdef DEBUG    
						cout << "Set to use device: " << optarg << endl;    
					#endif      
					break;
				case 'f':
					myFile = optarg;
					#ifdef DEBUG
						cout << "Set to send file: " << myFile << endl;
					#endif	
					break;
				case 's':
					programSlot = atoi(optarg);
					#ifdef DEBUG
						cout << "Program will reside at slot: " << programSlot << endl;
					#endif
					break;				
				case 't':
					mode_test = true; 
					break;		
				case 'v':
					version();
					return 0;
					break; 
				default:
					printf("Usage: bside -d [device] -f [filename]\n\n");
					break;
			}
		}
	}	
	else
		testRun = 1;

	/* set Alarm for 20 seconds */
	signal(SIGALRM,alarmTimeout);
	alarm(5);

	/* Setup the Stamp and try to identify it*/
	if (!myStamp.setup()){

		cout <<"Stamp could not be found or identified!\n";
		cout <<"If no reason is stated, Possible failure are:\n";
		cout <<"\t -Unsupported Stamp or firmware\n";
		cout <<"\t -Dead Battery on Board\n";
		cout <<"\t -Didn't specify proper device\n";
		return 1;
	}
	
	/* Disable the Alarm */
	alarm(0);

	cout  << "\nBasic Stamp detected!" << endl;
	cout  << "Model: "<< myStamp.get_model() << endl;
	cout  << "Vender: " << myStamp.get_vender() << endl;
	cout  << "Firmware version: " << setprecision(2) <<  myStamp.get_firmware() << endl;		

	if (!mode_test){
		/* If inFile was set, send it to the stamp */
		if( myFile.compare("") != 0){

			if (!myStamp.send_file(myFile, programSlot) ){

				cerr << "Transfering File Failed! (Reason Stated Above)\n";
				myStamp.halt();
				return 1;
			}

			cout << myFile << " transferred successfully!\n";
		}
	}
		
	/* Show the Debug Screen */
	stamp_debug();

	/* Kill everything */
	myStamp.halt();
	
	return 0;
}

/* Debug output  */
void stamp_debug(void)
{
	fd = myStamp.get_fd();
	char recvchar;

	myStamp.reset();

	cout << "Entering Debug Mode:\n";
	cout << "DEBUG OUTPUT:     (Press [Ctrl] + [c] to complete sequence)\n";
	cout << "_______________________________________________________________\n";

	signal(SIGINT,(sighandler_t)control_c_handler);
	debug_loop_done = 0;

	do
	{
    		fflush(stdout);

		if (read(fd,&recvchar, 1))
	  	{
	    		cout << recvchar;

	    		/* Take into account DOS EOL: */
	    		if (recvchar == '\r')
	      			cout << endl;

	    		fflush(stdout);
	  	}
      		
	}while (!debug_loop_done);
}

/* handle control c */
void *control_c_handler(int sig)
{
	fflush(stdout);
	fflush(stderr);
	
	cout << "_______________________________________________________________\n";
	cout << "\nUser Aborted\n";

	debug_loop_done = 1;
	return NULL;
}

/* Handle any timeout */
void alarmTimeout(int signal)
{
	cout << "Could not detect a stamp. Please check Your connections\n";
	exit(-13);
}
